File: scripts/PyMWIR/Signal.py
import torch
from . import PyBindMWIR

class Signal:
    def __init__(self, frequency_range = (1 * torch.pi * 1e9, 1 * torch.pi * 1e9), n_samples = 1, impl = None):
        if impl is None:
            frequency_range, n_samples = self.__SetFrequencyRange(frequency_range, n_samples)
            self.signal = PyBindMWIR.Signal(frequency_range, n_samples)
        elif type(impl) is PyBindMWIR.Signal:
            self.signal = impl
        else:
            raise TypeError("impl must be of type Signal or None")

    def SetFrequencyRange(self, frequency_range, n_samples):
        if self.signal is None:
            raise ValueError("internal signal instance has been moved.")
        frequency_range, n_samples = self.__SetFrequencyRange(frequency_range, n_samples)
        self.signal.SetFrequencyRange(frequency_range, n_samples)

    def GetFrequencyRange(self):
        if self.signal is None:
            raise ValueError("internal signal instance has been moved.")
        return self.signal.GetFrequencyRange()
    
    def GetNSamples(self):
        if self.signal is None:
            raise ValueError("internal signal instance has been moved.")
        return self.signal.GetNSamples()
    
    def GetFStep(self):
        if self.signal is None:
            raise ValueError("internal signal instance has been moved.")
        return self.signal.GetFStep()


    def __SetFrequencyRange(self, frequency_range, n_samples):
        if(type(frequency_range) is torch.Tensor):
            if(len(frequency_range.shape) != 1 or frequency_range.shape[0] != 2):
                raise ValueError("position must be a tensor of shape [2]")
        elif (type(frequency_range) is list or type(frequency_range) is tuple):
            if(len(frequency_range) != 2):
                raise ValueError("position must be a list or tuple of length 3")
        else:
            raise TypeError("position must be a torch.Tensor, list, or tuple")
        __frequency_range = torch.tensor(frequency_range, dtype=torch.float32)

        if(type(n_samples) is not int or n_samples <= 0):
            raise ValueError("n_samples must be a positive integer")
        __n_samples = torch.tensor((n_samples,), dtype=torch.int32)

        return __frequency_range, __n_samples

File: src/pybindmwir/signal.cpp
#include "pybindmwir/signal.hpp"

namespace py = pybind11;

Signal::Signal()
{
    mwir_signal_ = std::make_unique<MWIR::Signal>(std::nullopt, std::nullopt);
}

Signal::Signal(std::unique_ptr<MWIR::Signal> &&impl)
{
    if (!impl)
    {
        throw std::invalid_argument("Signal implementation cannot be null.");
    }

    mwir_signal_ = std::move(impl);
}

Signal::Signal(const torch::Tensor &frequency_range, const torch::Tensor &n_samples)
{
    mwir_signal_ = std::make_unique<MWIR::Signal>(glm::vec2(frequency_range[0].item<float>(), frequency_range[1].item<float>()), n_samples[0].item<int>());
}


Signal Signal::Clone() const
{
    if (!mwir_signal_)
    {
        throw std::runtime_error("Signal ownership has been transferred.");
    }

    return Signal(std::move(std::make_unique<MWIR::Signal>(mwir_signal_->Clone())));
}

void Signal::SetFrequencyRange(const torch::Tensor &frequency_range, const torch::Tensor &n_samples)
{
    if(!mwir_signal_)
    {
        throw std::runtime_error("Signal ownership has been transferred.");
    }

    mwir_signal_->SetFrequencyRange(glm::vec2(frequency_range[0].item<float>(), frequency_range[1].item<float>()), n_samples[0].item<int>());
}

torch::Tensor Signal::GetFrequencyRange() const
{
    if(!mwir_signal_)
    {
        throw std::runtime_error("Signal ownership has been transferred.");
    }

    glm::vec2 freq_range = mwir_signal_->GetFrequencyRange();
    return torch::tensor({freq_range.x, freq_range.y}, torch::kFloat32).view({2});
}

torch::Tensor Signal::GetNSamples() const
{
    if(!mwir_signal_)
    {
        throw std::runtime_error("Signal ownership has been transferred.");
    }

    return torch::tensor(mwir_signal_->GetNSamples(), torch::kInt32).view({1});
}

torch::Tensor Signal::GetFStep() const
{
    if(!mwir_signal_)
    {
        throw std::runtime_error("Signal ownership has been transferred.");
    }

    return torch::tensor(mwir_signal_->GetFStep(), torch::kFloat32).view({1});
}

void init_signal(py::module_ &m)
{
    py::class_<Signal, std::shared_ptr<Signal>>(m, "Signal")
        .def(py::init<>())
        .def(py::init<torch::Tensor&, torch::Tensor&>())
        .def("Clone", &Signal::Clone)
        .def("SetFrequencyRange", &Signal::SetFrequencyRange)
        .def("GetFrequencyRange", &Signal::GetFrequencyRange)
        .def("GetNSamples", &Signal::GetNSamples)
        .def("GetFStep", &Signal::GetFStep);
}